### PR TITLE
chore: nest test steps based on zones

### LINKED ISSUE #21423: [BUG] Html reporter doesn't show one of the steps.
### System info
- Playwright Version: v1.31.2
- Operating System: Windows 11
- Browser: Chromium
- Other info:

### Source code
- [x] I provided exact source code that allows reproducing the issue locally.

**Link to the GitHub repository with the repro**

https://github.com/Marcin3/challenge

or

**Config file**

```js
// playwright.config.ts
import type { PlaywrightTestConfig } from "@playwright/test";
import { devices } from "@playwright/test";
import dotenv from "dotenv";

dotenv.config();

const config: PlaywrightTestConfig = {
  testDir: "./tests",
  /* Maximum time one test can run for. */
  timeout: 30 * 1000,
  expect: {
    /**
     * Maximum time expect() should wait for the condition to be met.
     * For example in `await expect(locator).toHaveText();`
     */
    timeout: 10000,
  },
  fullyParallel: true,
  forbidOnly: !!process.env.CI,
  retries: process.env.CI ? 2 : 0,
  workers: 2,
  reporter: "html",

  use: {
    actionTimeout: 0,
    baseURL: process.env.BASE_URL,

    /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
    trace: "on",
  },

  /* Configure projects for major browsers */
  projects: [
    {
      name: "chromium",
      use: {
        ...devices["Desktop Chrome"],
      },
    },
  ],
};

export default config;

```

**Test file (self-contained)**

```ts
test("Search wrong word", async ({ page }) => {
  const wrongWord = "sythetic";
  const mainMenuPage = new MainMenuPage(page);

  await test.step(`Given User is on page https://mostly.ai/`, async () => {
    await page.goto("");
    expect(page).toHaveURL("");
  });

  await test.step(`When User click the “Search” button`, async () => {
    await mainMenuPage.clickOnSearchButton();
  });

  await test.step(`And type “sythetic” (wrong spelling of synthetic) in search field`, async () => {
    await mainMenuPage.fillSearchFiled(wrongWord);
  });

  await test.step(`And press Enter`, async () => {
    await mainMenuPage.pressEnterButtonOnSearchFiled();
  });

  await test.step(`Then User can see following information: “Sorry, no results for: sythetic”`, async () => {
    const firstPart = `Sorry, no results for:`;
    await mainMenuPage.checkSearchResult(firstPart, wrongWord);
  });
});
```

**Steps**
- Clone the repo
- First install project using command ` npm install`
- To run tests please run from root using command `npx playwright test`
- Open report  npx playwright show-report

**Expected**

For test Search wrong word you should see step When User click the “Search” button

**Actual**
Now this step is missing 
image

Here is also link to playwright report https://github.com/Marcin3/challenge/tree/main/playwright-report

### PR DESCRIPTION
**Pull request by pavelfeldman**
Fixes: #21423,

### CODE REVIEW

File: packages/playwright-test/src/worker/testInfo.ts [hunk 1]
```diff
@@ -292,16 +300,18 @@ export class TestInfoImpl implements TestInfo {
     this._hasHardError = state.hasHardError;
   }
 
-  async _runAsStep<T>(cb: () => Promise<T>, stepInfo: Omit<TestStepInternal, 'complete' | 'wallTime'>): Promise<T> {
+  async _runAsStep<T>(cb: (ste: TestStepInternal) => Promise<T>, stepInfo: Omit<TestStepInternal, 'complete' | 'wallTime' | 'parentStepId' | 'stepId'>): Promise<T> {
```
Comment (yury-s) [inline packages/playwright-test/src/worker/testInfo.ts:303]:
```suggestion
  async _runAsStep<T>(cb: (step: TestStepInternal) => Promise<T>, stepInfo: Omit<TestStepInternal, 'complete' | 'wallTime' | 'parentStepId' | 'stepId'>): Promise<T> {
```

File: packages/playwright-test/src/worker/workerMain.ts [hunk 2]
```diff
@@ -386,100 +382,96 @@ export class WorkerMain extends ProcessRunner {
...
+          // Attribute to 'teardown' because worker fixtures are not perceived as a part of a test.
+          testInfo._timeoutManager.setCurrentRunnable({ type: 'teardown', slot: teardownSlot });
+          debugTest(`tearing down worker scope started`);
+          const workerScopeError = await testInfo._runFn(() => this._fixtureRunner.teardownScope('worker', testInfo._timeoutManager));
+          debugTest(`tearing down worker scope finished`);
+          firstAfterHooksError = firstAfterHooksError || workerScopeError;
+        });
+      }
+      if (firstAfterHooksError)
+        step.complete({ error: firstAfterHooksError });
```
Comment (yury-s) [inline packages/playwright-test/src/worker/workerMain.ts:469]: testInfo._runAsStep will call step.complete({}) again, is that expected?

**Review by yury-s — APPROVED** [review]

File: packages/playwright-test/src/worker/workerMain.ts [hunk 3]
```diff
@@ -320,17 +320,10 @@ export class WorkerMain extends ProcessRunner {
...
-      const beforeHooksStep = testInfo._addStep({
-        category: 'hook',
-        title: 'Before Hooks',
-        canHaveChildren: true,
-        forceNoParent: true,
-        wallTime: Date.now(),
-      });
-
+      let params: object | null = null;
       // Note: wrap all preparation steps together, because failure/skip in any of them
```
Comment (dgozman) [inline packages/playwright-test/src/worker/workerMain.ts:324]: This comment suggests that running `beforeEach` from `test` in separate `runFn()` calls will break "skip inside beforeEach" scenario. I guess we are missing a test for this!

File: packages/playwright-test/src/worker/workerMain.ts (hunk 2 above)
Comment (pavelfeldman) [reply packages/playwright-test/src/worker/workerMain.ts:469]: Yes, that is alright.

**Review by dgozman — APPROVED** [review]

**Pull request merged**